### PR TITLE
Allow anonymous user access on update hook for flexible groups

### PR DESCRIPTION
**Pull request by NickDaelemans**
## Problem
When viewing a public group, anonymous users don't see the album link, although the module is enabled.

## Solution
By adding the right permissions to the anonymous users for flexible groups, anonymous users are able to see both the Albums link and Albums page.

## Issue tracker
https://www.drupal.org/project/social/issues/3309222#comment-14689753

## How to test
- [ ] Make sure the `social_album` module is installed
- [ ] Create a public group
- [ ] As an anonymous user
- [ ] Open the public group
- [ ] You should see the Albums link
- [ ] Click on the Albums link
- [ ] You should see `No albums in this group`
- [ ] If you add a public album, this should also be visible when viewing the Albums page.

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] New features or changes to existing features are covered by tests, either unit (preferably) or behat
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Change Record
Please be aware that this change has permission changes. This is going to change the way users are able to view albums.

AN users will be able to view any public album you create
LU users will be able to view any public and community album you create.

### CODE REVIEW

File: modules/social_features/social_album/social_album.install [hunk 1]
```diff
@@ -307,3 +307,12 @@ function social_album_update_11401(): string {
...
   return $updateHelper->logger()->output();
 }
+
+/**
+ * Grant albums permissions for anonymous users.
+ */
+function social_album_update_11402(): void {
+  /** @var \Drupal\group\Entity\GroupRoleInterface $group_role_anonymous */
+  $group_role_anonymous = \Drupal::entityTypeManager()->getStorage('group_role')->load('flexible_group-anonymous');
+  $group_role_anonymous->grantPermissions(['view group_node:album content', 'view group_node:album entity'])->save();
```
[the remark this answers is not in the document]
Comment (NickDaelemans) [reply modules/social_features/social_album/social_album.install:317]: I do think this needs some proper testing before merging. To answer your questions:

- The group permissions for anonymous users are indeed not available in UI. Hence the `permissions_ui: false` in the `group.role.flexible_group-anonymous.yml` file.
- I think you might be right; what this update should do, is only show the `Album overview` page. However, the view that builds that page uses the `view group_node:album content`, that's why I added it here.
- I think we should indeed add (or replace with) `View node.album.field_content_visibility:public content`. This is the one I enabled on the client to fix the issue that they could not see a single public Album. 

I will work further on this, to investigate the good remarks you made here, thanks!

**Comment by tbsiqueira** [comment]
Hi @NickDaelemans , thank you for your contribution! 

Can you please create a ticket on d.org for this PR? Also, please add release notes and screenshots of the expected behavieur.

**Review by tbsiqueira — APPROVED** [review]

**Review by tbsiqueira — CHANGES_REQUESTED** [review]
If there aren't, please add behat tests to cover this change.

- Create album as community content;
- Create algum as group content;
- Create album as public content;
- Create groups with the respective permissions;
- Check if LU and AN users have the right permissions

File: modules/social_features/social_album/social_album.install [hunk 2]
```diff
@@ -307,3 +324,51 @@ function social_album_update_11401(): string {
...
+    RoleInterface::ANONYMOUS_ID,
+    [
+      'view node.album.field_content_visibility:public content',
+    ]
+  );
+
+  user_role_grant_permissions(
+    'verified',
+    [
+      'view node.album.field_content_visibility:community content',
```
Comment (navneet0693) [inline modules/social_features/social_album/social_album.install:364]: I see that this was already added in https://github.com/goalgorilla/open_social/blob/main/modules/social_features/social_album/social_album.install#L28, did few a reason why it was missed being added ?

File: modules/social_features/social_album/social_album.install [hunk 3]
```diff
@@ -307,3 +324,51 @@ function social_album_update_11401(): string {
...
+  /** @var \Drupal\group\Entity\GroupRoleInterface $group_role_anonymous */
+  $group_role_anonymous = \Drupal::entityTypeManager()->getStorage('group_role')->load('flexible_group-anonymous');
+  $group_role_anonymous->grantPermissions(['view group_node:album content', 'view group_node:album entity'])->save();
+}
+
+/**
+ * Grant group albums permissions for group outsiders.
+ */
+function social_album_update_11403(): void {
+  /** @var \Drupal\group\Entity\GroupRoleInterface $group_role_outsider */
```
Comment (navneet0693) [inline modules/social_features/social_album/social_album.install:341]: I think we can remove it, if we call `_social_album_set_group_permissions` and changes are added to that function.

**Review by navneet0693 — CHANGES_REQUESTED** [review]

**Comment by NickDaelemans** [comment]
Since we are not doing the Behat tests here, I created a new PR here: https://github.com/goalgorilla/open_social/pull/3121

**Pull request closed without merging**